File: frontend/js/views/macrs_depreciation.js
// MACRS depreciation — IRS tax depreciation schedule, via /calc/macrs-depreciation.
import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
export async function renderMacrsDepreciation(mount, _s) {
    const tok = currentViewToken(); if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.macrs.h1.title">// MACRS DEPRECIATION</span></h1>
        <p class="muted small" data-i18n="view.macrs.hint.intro">The U.S. Modified Accelerated Cost Recovery System tax depreciation schedule (GDS, half-year convention, IRS Pub. 946). Applies the published per-year percentage table for the recovery period to give the deduction, accumulated depreciation, and remaining basis each year. Not tax advice.</p>
        <div class="lpv-split"><div class="chart-panel"><h2 data-i18n="view.macrs.h2.inputs">Asset</h2>
        <form id="macrs-form" class="inline-form">
            <label><span data-i18n="view.macrs.label.label">Asset</span><input type="text" name="asset_label" value="Equipment"></label>
            <label><span data-i18n="view.macrs.label.basis">Depreciable basis ($)</span><input type="number" step="100" min="0" name="basis_usd" value="10000" required></label>
            <label><span data-i18n="view.macrs.label.period">Recovery period</span><select name="recovery_years">
                <option value="3">3-year</option><option value="5" selected>5-year</option><option value="7">7-year</option>
                <option value="10">10-year</option><option value="15">15-year</option><option value="20">20-year</option></select></label>
            <label><span data-i18n="view.macrs.label.year">Placed in service (year)</span><input type="number" step="1" name="placed_in_service_year" value="2026"></label>
        </form></div><div id="macrs-result" class="chart-panel lpv-preview"></div></div>`;
    applyUiI18n(mount);
    const form = mount.querySelector('#macrs-form');
    const gen = async () => {
        const body = { asset_label: (form.querySelector('[name="asset_label"]').value || '').trim(), basis_usd: Number(form.querySelector('[name="basis_usd"]').value) || 0, recovery_years: Number(form.querySelector('[name="recovery_years"]').value) || 5, placed_in_service_year: Number(form.querySelector('[name="placed_in_service_year"]').value) || 0 };
        try { const d = await api.calcMacrsDepreciation(body); if (!viewIsCurrent(tok)) return; res(mount, d, body.placed_in_service_year); }
        catch (e) { showToast(e.message || t('view.macrs.toast.error'), { level: 'error' }); }
    };
    const live = debounce(gen, 250); form.addEventListener('input', () => live()); form.addEventListener('submit', (e) => { e.preventDefault(); gen(); }); gen();
}
function res(mount, d, startYear) {
    const el = mount.querySelector('#macrs-result');
    if (!d.valid) { el.innerHTML = `<p class="muted" data-i18n="view.macrs.invalid">Recovery period must be 3, 5, 7, 10, 15, or 20 years.</p>`; applyUiI18n(el); return; }
    const rows = d.schedule.map((r) => `<tr><td>${startYear > 0 ? startYear + r.year - 1 : r.year}</td><td>${r.rate_pct}%</td><td>${money(r.depreciation_usd)}</td><td>${money(r.accumulated_usd)}</td><td>${money(r.remaining_basis_usd)}</td></tr>`).join('');
    el.innerHTML = `<div class="lpv-bar"><div class="cards">
        <div class="card pos"><div class="label" data-i18n="view.macrs.card.total">Total depreciation</div><div class="value">${money(d.total_depreciation_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.macrs.card.period">Recovery period</div><div class="value">${d.recovery_years}y</div></div>
    </div></div>
    <table class="data-table"><thead><tr><th data-i18n="view.macrs.th.year">Year</th><th data-i18n="view.macrs.th.rate">Rate</th><th data-i18n="view.macrs.th.dep">Depreciation</th><th data-i18n="view.macrs.th.accum">Accumulated</th><th data-i18n="view.macrs.th.basis">Remaining basis</th></tr></thead><tbody>${rows}</tbody></table>`;
    applyUiI18n(el);
}

File: crates/traderview-core/src/macrs_depreciation.rs
//! MACRS depreciation — the U.S. Modified Accelerated Cost Recovery System tax
//! depreciation schedule (GDS, half-year convention, IRS Pub. 946). Given the
//! depreciable basis and a recovery period (3, 5, 7, 10, 15, or 20 years), it
//! applies the published per-year percentage table to produce the deduction,
//! accumulated depreciation, and remaining basis for each year. Distinct from the
//! book depreciation schedule (straight-line / declining-balance) and the §1250
//! recapture module — this is the tax schedule. Not tax advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct MacrsInput {
    pub asset_label: String,
    /// Depreciable basis (cost).
    pub basis_usd: f64,
    /// Recovery period in years: 3, 5, 7, 10, 15, or 20.
    pub recovery_years: u32,
    #[serde(default)]
    pub placed_in_service_year: i32,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct MacrsRow {
    pub year: u32,
    pub rate_pct: f64,
    pub depreciation_usd: f64,
    pub accumulated_usd: f64,
    pub remaining_basis_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct MacrsReport {
    pub recovery_years: u32,
    pub total_depreciation_usd: f64,
    pub schedule: Vec<MacrsRow>,
    pub valid: bool,
}

/// GDS half-year-convention percentages by recovery period (IRS Pub. 946).
fn table(years: u32) -> Option<&'static [f64]> {
    match years {
        3 => Some(&[33.33, 44.45, 14.81, 7.41]),
        5 => Some(&[20.00, 32.00, 19.20, 11.52, 11.52, 5.76]),
        7 => Some(&[14.29, 24.49, 17.49, 12.49, 8.93, 8.92, 8.93, 4.46]),
        10 => Some(&[10.00, 18.00, 14.40, 11.52, 9.22, 7.37, 6.55, 6.55, 6.56, 6.55, 3.28]),
        15 => Some(&[
            5.00, 9.50, 8.55, 7.70, 6.93, 6.23, 5.90, 5.90, 5.91, 5.90, 5.91, 5.90, 5.91, 5.90,
            5.91, 2.95,
        ]),
        20 => Some(&[
            3.750, 7.219, 6.677, 6.177, 5.713, 5.285, 4.888, 4.522, 4.462, 4.461, 4.462, 4.461,
            4.462, 4.461, 4.462, 4.461, 4.462, 4.461, 4.462, 4.461, 2.231,
        ]),
        _ => None,
    }
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

pub fn generate(i: &MacrsInput) -> MacrsReport {
    let rates = match table(i.recovery_years) {
        Some(r) if i.basis_usd > 0.0 => r,
        _ => return MacrsReport { recovery_years: i.recovery_years, ..Default::default() },
    };
    let mut accumulated = 0.0;
    let schedule: Vec<MacrsRow> = rates
        .iter()
        .enumerate()
        .map(|(k, &rate)| {
            let dep = cents(i.basis_usd * rate / 100.0);
            accumulated = cents(accumulated + dep);
            MacrsRow {
                year: (k + 1) as u32,
                rate_pct: rate,
                depreciation_usd: dep,
                accumulated_usd: accumulated,
                remaining_basis_usd: cents(i.basis_usd - accumulated),
            }
        })
        .collect();

    MacrsReport {
        recovery_years: i.recovery_years,
        total_depreciation_usd: cents(accumulated),
        schedule,
        valid: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> MacrsInput {
        MacrsInput {
            asset_label: "Equipment".into(),
            basis_usd: 10_000.0,
            recovery_years: 5,
            placed_in_service_year: 2026,
        }
    }

    #[test]
    fn five_year_schedule() {
        let d = generate(&base());
        assert!(d.valid);
        assert_eq!(d.schedule.len(), 6);
        assert!(close(d.schedule[0].depreciation_usd, 2_000.0));
        assert!(close(d.schedule[1].depreciation_usd, 3_200.0));
        assert!(close(d.schedule[5].depreciation_usd, 576.0));
        assert!(close(d.total_depreciation_usd, 10_000.0));
    }

    #[test]
    fn accumulates_to_basis() {
        for yrs in [3, 5, 7, 10, 15, 20] {
            let d = generate(&MacrsInput { recovery_years: yrs, ..base() });
            assert!(close(d.total_depreciation_usd, 10_000.0), "yrs {yrs}");
            assert!(close(d.schedule.last().unwrap().remaining_basis_usd, 0.0));
        }
    }

    #[test]
    fn remaining_basis_declines() {
        let d = generate(&base());
        for w in d.schedule.windows(2) {
            assert!(w[1].remaining_basis_usd <= w[0].remaining_basis_usd);
        }
    }

    #[test]
    fn unsupported_period_invalid() {
        let d = generate(&MacrsInput { recovery_years: 4, ..base() });
        assert!(!d.valid);
    }

    #[test]
    fn zero_basis_invalid() {
        let d = generate(&MacrsInput { basis_usd: 0.0, ..base() });
        assert!(!d.valid);
    }
}
